modules/gateways/Blockonomics/Blockonomics.php

```diff
@@ -318,63 +372,243 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
-			"timestamp"=> $existing_order->timestamp,
-			"status" => $existing_order->status,
-			"value" => $existing_order->value,
-			"bits" => $existing_order->bits,
-			"bits_payed" => $existing_order->bits_payed
-		);
+		return $first_order->id_order;
+	}
 
-		return $row_in_array;
+	/*
+	 * Get all orders linked to uuid
+	 */	
+	public function getAllOrdersByHash($order_hash) {
+		$order_id = $this->decrypt($order_hash);
+		try {
+			return Capsule::table('blockonomics_bitcoin_orders')
+				->where('id_order', $order_id)
+				->orderBy('timestamp', 'desc')->get();
+		} catch (\Exception $e) {
+				exit("Unable to get orders from blockonomics_bitcoin_orders: {$e->getMessage()}");
+		}
+	}
+
+	/*
+	 * Check for pending orders and return if exists
+	 */	
+	public function getPendingOrder($orders) {
+		foreach ($orders as $order) {
+			//check if status 0 or 1
+			if($order->status == 0 || $order->status == 1){
+				return $order;
+			}
+		}
+	}
+
+	/*
+	 * Check for unused address for the currency linked to the order
+	 */		
+	public function getAndUpdateWaitingOrder($orders, $blockonomics_currency) {
+		foreach ($orders as $order) {
+			//check for currency address already waiting
+			if($order->blockonomics_currency == $blockonomics_currency && $order->status == -1){
+				//Check if existing order is expired
+				$current_time = time();
+				$total_time = $this->getTimePeriod() * 60;
+				$clock = $order->timestamp + $total_time - $current_time;
+				if($clock < 0){
+					$order->bits = $this->convertFiatToBlockonomicsCurrency($order->value, $order->blockonomics_currency);
+					$order->timestamp = $current_time;
+				}
+				$this->updateOrderExpected($order->addr, $order->blockonomics_currency, $order->timestamp, $order->bits);
+				return $order;
+			}
+		}
+	}
+
+	/*
+	 * Check for unused address or create new
+	 */	
+	public function processNewOrder($orders, $blockonomics_currency) {
+		foreach ($orders as $order) {
+			//check for new blank order in db
+			if($order->addr == ""){
+				$new_addresss_response = $this->getNewAddress($blockonomics_currency);
+				if ($new_addresss_response->response_code == 200){
+					$order->addr = $new_addresss_response->address;
+				}else{
+					exit($new_addresss_response->message);
+				}
+
+				$this->updateOrderAddress($order->id_order, $order->addr, $blockonomics_currency);
+				$order->blockonomics_currency = $blockonomics_currency;
+				$order->bits = $this->convertFiatToBlockonomicsCurrency($order->value, $order->blockonomics_currency);
+				$order->timestamp = time();
+				$this->updateOrderExpected($order->addr, $order->blockonomics_currency, $order->timestamp, $order->bits);
+				return $order;
+			}
+		}
+		// blank order already used, create a new blank order with same uuid
+		// fix this
+		$order_uuid = $this->newOrder($orders[0]->value, $orders[0]->id_order);
+		$order = $this->processOrderHash($order_uuid, $blockonomics_currency);
+		return $order;
+	}
+
```

Comment:
I am not sure why these lines are needed. This creates a potentially recursive loop since processOrderHash calls this function again

Reply:
These lines are required when the user has generated an address already, but did not pay, then selects another currency. There is no skeleton order waiting. Updated to prevent possible loop issues in 96d9dd4